parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java

```diff
@@ -36,42 +36,152 @@
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        if (params.size() != 2) {
+          throw new RuntimeException("Expecting 2 parameters for time logical type, got " + params.size());
+        }
+        return timeType(Boolean.parseBoolean(params.get(1)), TimeUnit.valueOf(params.get(0)));
+      }
+    },
+    TIMESTAMP {
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        if (params.size() != 2) {
+          throw new RuntimeException("Expecting 2 parameters for timestamp logical type, got " + params.size());
+        }
+        return timestampType(Boolean.parseBoolean(params.get(1)), TimeUnit.valueOf(params.get(0)));
+      }
+    },
+    INT {
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        if (params.size() != 2) {
+          throw new RuntimeException("Expecting 2 parameters for integer logical type, got " + params.size());
+        }
+        return intType(Integer.valueOf(params.get(0)), Boolean.parseBoolean(params.get(1)));
+      }
+    },
+    JSON {
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        return jsonType();
+      }
+    },
+    BSON {
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        return bsonType();
+      }
+    },
+    INTERVAL {
+      @Override
+      protected LogicalTypeAnnotation fromString(List<String> params) {
+        return IntervalLogicalTypeAnnotation.getInstance();
+      }
+    };
+
+    protected abstract LogicalTypeAnnotation fromString(List<String> params);
+  }
+
   /**
    * Convert this parquet-mr logical type to parquet-format LogicalType.
    *
    * @return the parquet-format LogicalType representation of this logical type implementation
    */
-  LogicalType toLogicalType();
+  public abstract LogicalType toLogicalType();
 
   /**
    * Convert this parquet-mr logical type to parquet-format ConvertedType.
    *
    * @return the parquet-format ConvertedType representation of this logical type implementation
    */
-  ConvertedType toConvertedType();
+  public abstract ConvertedType toConvertedType();
 
   /**
    * Convert this logical type to old logical type representation in parquet-mr (if there's any).
    * Those logical type implementations, which don't have a corresponding mapping should return null.
    *
    * @return the OriginalType representation of the new logical type, or null if there's none
    */
-  OriginalType toOriginalType();
+  public abstract OriginalType toOriginalType();
 
   /**
    * Visits this logical type with the given visitor
    *
    * @param logicalTypeAnnotationVisitor the visitor to visit this type
    */
-  void accept(LogicalTypeAnnotationVisitor logicalTypeAnnotationVisitor);
+  public abstract void accept(LogicalTypeAnnotationVisitor logicalTypeAnnotationVisitor);
+
+  public abstract LogicalTypes getType();
```

Comment:
It seems this one is not used outside of this class/package. I would suggest not being public.

Reply:
accept is intended for public use, for the visitors

Reply:
Sorry, I meant `getType()`.